client/backoff/backoff.go

```diff
@@ -52,3 +53,18 @@ func (b *Backoff) Duration() time.Duration {
 func (b *Backoff) Reset() {
 	b.x = 0
 }
+
+// Retry calls a function until it succeeds or the context is done.
+// It will backoff if the function returns an error.
+// Once the context is done, retries will end and the most recent error will be returned.
+// Backoff is not reset by this function.
+func (b *Backoff) Retry(ctx context.Context, f func() error) error {
+	var err error
+	for ctx.Err() == nil {
+		if err = f(); err == nil {
+			break
+		}
+		time.Sleep(b.Duration())
```

Comment:
Just a thought - duration could grow to be quite large, and if someone cancels the `ctx` they could end up still waiting quite a while for this to give up and return.
Could do something along the lines of:
```go
select {
case _ = <-time.After(Duration()):
	continue
case _ = <-ctx.Done():
	return lastErr
}
```
Quick demo: https://play.golang.org/p/kw3UMmr7YE